### PR TITLE
[4.0] Cleanup error handling in the installer on failed DB Connections

### PR DESCRIPTION
**Pull request by wilsonge**
You can test the different kinds of errors. All errors messages now use the JFactory enqueued messages rather than a extra message field. Technically we now handle the case the server spits out an exception (`error` is `true`) although I don't know how this would happen.

Finally we now get errors specific to the db property failing being rendered. You can test the following things:

1. Enter an invalid DB prefix (add an invalid character like an exclamation mark) - you'll get the `Check your DB credentials, db type, db name or hostname` string for form validation failure as well as the specific db prefix string (I've restored the Joomla 3 one which seemed to have been deleted along the way despite it's being used in 3 different places)

2. Enter an invalid password. The error message will show up with a more specific error than currently

e.g. 

```
Could not connect to the database. Connector returned number: Could not connect to MySQL: Access denied for user 'root'@'localhost' (using password: YES)
```

(yes the message is kinda grim reading but this is the `INSTL_DATABASE_COULD_NOT_CONNECT` string - this can be refined in the future for sure :)

3. Obviously test the happy installer path continues to work

### CODE REVIEW

File: installation/language/en-GB/en-GB.ini [hunk 1]
```diff
@@ -39,6 +39,7 @@ INSTL_DATABASE_NAME_LABEL="Database Name"
 INSTL_DATABASE_NO_SCHEMA="No database schema exists for this database type."
 INSTL_DATABASE_PASSWORD_DESC="Either a password you created or a password provided by your host."
 INSTL_DATABASE_PREFIX_DESC="Enter a table prefix or use the randomly generated one."
+INSTL_DATABASE_PREFIX_MSG="The table prefix must start with a letter, be followed by optional alphanumeric characters and by an underscore"
```
Comment (brianteeman) [inline installation/language/en-GB/en-GB.ini]: optionally be followed by alpha....

File: installation/src/Controller/InstallationController.php [hunk 2]
```diff
@@ -60,12 +60,16 @@ public function dbcheck()
...
 		$model = $this->getModel('Setup');
-		if ($model->checkForm('setup') === false || $model->validateDbConnection() === false)
+		if ($model->checkForm('setup') === false)
 		{
-			$r->messages = Text::_('INSTL_DATABASE_VALIDATION_ERROR');
-			$r->view = 'setup';
+			$this->app->enqueueMessage(Text::_('INSTL_DATABASE_VALIDATION_ERROR'), 'error');
+			$r->validated = false;
+			$this->sendJsonResponse($r);
+			return;
```
Comment (Quy) [inline installation/src/Controller/InstallationController.php]: Add blank line above `return;`

File: installation/language/en-GB/en-GB.ini (hunk 1 above)
Comment (wilsonge) [reply installation/language/en-GB/en-GB.ini]: Done although still doesn't quite read right to me :/

File: installation/template/js/setup.js [hunk 3]
```diff
@@ -84,20 +84,21 @@ Joomla.checkDbCredentials = function() {
...
-			} else {
-				Joomla.loadingLayer('hide');
+			Joomla.loadingLayer("hide");
 
-				// Run the installer - we let this handle the redirect for now
-				// TODO: Convert to promises
-				Joomla.install(['config'], form);
+			if (response.error) {
+				Joomla.renderMessages({'error': [response.message]});
+			} else {
```
Comment (C-Lodder) [inline installation/template/js/setup.js]: `else if (foo) { ... }` === `else { if (foo) {  ... } }`

Comment (wilsonge) [reply installation/template/js/setup.js]: Nice catch :)

**Review by brianteeman — APPROVED** [review]

File: installation/language/en-GB/en-GB.ini (hunk 1 above)
Comment (brianteeman) [reply installation/language/en-GB/en-GB.ini]: Took a fresh look

+INSTL_DATABASE_PREFIX_MSG="The table prefix must start with a letter and end with an underscore. Optionally additional alphanumeric characters can be included."

**Comment by laoneo** [comment]
Worked as expected. Thanks.

**Comment by wilsonge** [comment]
Thanks!

**Comment by dgrammatiko** [comment]
@wilsonge @laoneo can we apply the same idea for all errors in the installation app ?

**Comment by wilsonge** [comment]
Probably - my main aim here was to remove the messages property - which isn't used by the rest of the `InstallationController` https://github.com/joomla/joomla-cms/blob/4.0-dev/installation/src/Controller/InstallationController.php

**Comment by laoneo** [comment]
Do we need first the promise based set up?

**Comment by dgrammatiko** [comment]
@laoneo good question, let me finish the monkey work with the es6 scripts and then I 'll spend some time on this "promise land"...

**Pull request merged**